### DISCUSSION TITLE
🧪 testax: Jit-able runtime assertions for JAX in NumPy style.

### DISCUSSION THREAD

**Discussion by tillahoffmann** [jax-ml/jax#20039] [Show and tell]
[![](https://img.shields.io/pypi/v/testax)](https://pypi.org/project/testax) [![](https://github.com/tillahoffmann/testax/actions/workflows/build.yml/badge.svg)](https://github.com/tillahoffmann/testax/actions/workflows/build.yml) [![](https://readthedocs.org/projects/testax/badge/?version=latest)](https://testax.readthedocs.io/en/latest/?badge=latest)

[testax](https://github.com/tillahoffmann/testax) provides `jit`-able runtime assertions for JAX through the [testing interface](https://numpy.org/doc/stable/reference/routines.testing.html) familiar to NumPy users.

# Overview

```python
>>> import jax
>>> from jax import numpy as jnp
>>> import testax
>>>
>>> def safe_log(x):
...     testax.assert_array_less(0, x)
...     return jnp.log(x)
>>>
>>> safe_log(jnp.arange(2))
Traceback (most recent call last):
    ...
jax._src.checkify.JaxRuntimeError:
Arrays are not less-ordered

Mismatched elements: 1 / 2 (50%)
Max absolute difference: 1
Max relative difference: 1
 x: Array(0, dtype=int32, weak_type=True)
 y: Array([0, 1], dtype=int32)
```

testax assertions are `jit`-able, although errors need to be functionalized to conform to JAX’s requirement that functions are pure and do not have side effects (see the checkify [guide](https://jax.readthedocs.io/en/latest/debugging/checkify_guide.html) for details). In short, a `checkify`-d function returns a tuple `(error, value)`. The first element is an error that may have occurred, and the second is the return value of the original function.

```python
>>> jitted = jax.jit(safe_log)
>>> checkified = testax.checkify(jitted)
>>> error, y = checkified(jnp.arange(2))
>>> error.throw()
Traceback (most recent call last):
    ...
jax._src.checkify.JaxRuntimeError:
Arrays are not less-ordered

Mismatched elements: 1 / 2 (50%)
Max absolute difference: 1
Max relative difference: 1
 x: Array(0, dtype=int32, weak_type=True)
 y: Array([0, 1], dtype=int32)
>>> y
Array([-inf,   0.], dtype=float32)
```

**Comment by patrick-kidger** [comment]
[`equinox.error_if`](https://docs.kidger.site/equinox/api/errors/) may also be of interest here.

**Reply by tillahoffmann** [reply]
Neat! Is there an option to apply the string formatting only once an error is raised? E.g., using the current implementation, all the formatting happens in [`__str__`](https://github.com/tillahoffmann/testax/blob/9d1d84334cd56dac43cbfcc08a7e2576cf3313a5/testax/__init__.py#L82) of the exception class which is only called if an error occurred.

**Reply by patrick-kidger** [reply]
Not right now, but it's come up before, and I'd be happy to take a PR adding this: https://github.com/patrick-kidger/equinox/issues/644
